components/tidb_query_expr/src/impl_string.rs

```diff
@@ -1096,6 +1132,74 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_locate_2_args_utf8() {
+        let cases = vec![
+            // normal cases
+            (Some("bar"), Some("foobarbar"), Some(4i64)),
+            (Some("xbar"), Some("foobar"), Some(0i64)),
+            (Some(""), Some("foobar"), Some(1i64)),
+            (Some("foobar"), Some(""), Some(0i64)),
+            (Some(""), Some(""), Some(1i64)),
+            (Some("好世"), Some("你好世界"), Some(2i64)),
+            (Some("界面"), Some("你好世界"), Some(0i64)),
+            (Some("b"), Some("中a英b文"), Some(4i64)),
+            (Some("BaR"), Some("foobArbar"), Some(4i64)),
+            // null cases
+            (None, Some(""), None),
+            (None, Some("foobar"), None),
+            (Some(""), None, None),
+            (Some("foobar"), None, None),
+            (None, None, None),
+        ];
+
+        for (substr, s, exp) in cases {
+            let output = RpnFnScalarEvaluator::new()
+                .push_param(substr.map(|substr| substr.as_bytes().to_vec()))
+                .push_param(s.map(|s| s.as_bytes().to_vec()))
+                .evaluate(ScalarFuncSig::Locate2ArgsUtf8)
+                .unwrap();
+            assert_eq!(output, exp);
+        }
+    }
+
+    #[test]
+    fn test_locate_3_args_utf8() {
+        let cases = vec![
```

Comment:
Add invalid utf8 test cases

Reply:
done

BTW, to get ride of unnecessary error-handling, `-1i64` is imported as an error value. Any better implementation?